src/main/java/org/opensearch/searchrelevance/experiment/PointwiseExperimentProcessor.java

```
@@ -0,0 +1,254 @@
+            processExperimentWithJudgments(
+                experimentId,
+                queryText,
+                indexAndQueries,
+                judgmentList,
+                size,
+                docIdToScores,
+                hasFailure,
+                listener
+            );
+        }).exceptionally(e -> {
+            if (hasFailure.compareAndSet(false, true)) {
+                listener.onFailure(new Exception("Failed to load judgments", e));
+            }
+            return null;
+        });
+    }
+
+    /**
+     * Load and cache judgments for the experiment
+     */
+    private CompletableFuture<Map<String, String>> loadJudgmentsAsync(String experimentId, List<String> judgmentList, String queryText) {
+        // Check cache first
+        String cacheKey = experimentId + ":" + queryText;
+        if (judgmentCache.containsKey(cacheKey)) {
+            return CompletableFuture.completedFuture(judgmentCache.get(cacheKey));
+        }
+
+        // Load judgments in parallel
+        List<CompletableFuture<SearchResponse>> judgmentFutures = judgmentList.stream().map(judgmentId -> {
+            CompletableFuture<SearchResponse> future = new CompletableFuture<>();
+            judgmentDao.getJudgment(judgmentId, ActionListener.wrap(future::complete, future::completeExceptionally));
+            return future;
+        }).toList();
+
+        return CompletableFuture.allOf(judgmentFutures.toArray(new CompletableFuture[0])).thenApply(v -> {
+            Map<String, String> docIdToScores = new HashMap<>();
+
+            for (CompletableFuture<SearchResponse> future : judgmentFutures) {
+                try {
+                    SearchResponse response = future.join();
+                    extractJudgmentScores(queryText, response, docIdToScores);
+                } catch (Exception e) {
+                    log.error("Failed to process judgment response: {}", e.getMessage());
```

the exception will be swallowed here. Is it expected that we don't care about failed judgments reading ?

---

My reasoning is - experiment may be large, so if we fail for one judgment we still can continue the rest of experiment. I can do more sophisticated logic - count each failure and if we fail min_of(number of tasks, 5) times, then fail whole experiment, otherwise ignore failures